### ISSUE TITLE
enhancement request: Enable concurrency for PUT file with presigned url (dictionary changed size during iteration)

### ISSUE THREAD

**Issue by ArzelaAscoIi** [localstack/localstack#8011]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Enhancement description

We are using localstack for generating a bunch of presigned URLs and uploading files. These requests are sent concurrently, but localstack raises an exception `dictionary changed size during iteration`. I guess this happens because the webserver uses a shared dict for these concurrent requests. 

It would be helpful if localstack would support basic concurrency here :) Thanks in advance!

### 🧑‍💻 Implementation

- use async/await without sharing objects between requests

### Anything else?

I could have a look at this if you help me by pointing me to the right paces :)

**Comment by bentsku** [comment]
Hello @ArzelaAscoIi and thanks for your report.

Could you please share a bit more about the issue you're encountering? What version of LocalStack are you using? We've just rewritten our S3 implementation and it will the default in the next major release planned today, and already available in `latest`. Also, could you share the full traceback or the error you encountered?

We support concurrency in our webserver, but there could be an issue later in the chain with S3. You can have more verbose logs by setting `DEBUG=1` and `LS_LOG=trace`, but if there's a Python exception, the traceback should be enough.

Thanks!

**Comment by ArzelaAscoIi** [comment]
Hi @bentsku. We are using the image `localstack/localstack:1.4.0`. This is the stacktrace we got when we enable concurrency (even starts with 3 concurrent requests). Happy to support more information if required :)

```
Traceback (most recent call last):
  File "/opt/code/localstack/localstack/aws/chain.py", line 90, in handle
    handler(self, self.context, response)
  File "/opt/code/localstack/localstack/aws/handlers/service.py", line 122, in __call__
    handler(chain, context, response)
  File "/opt/code/localstack/localstack/aws/handlers/service.py", line 92, in __call__
    skeleton_response = self.skeleton.invoke(context)
  File "/opt/code/localstack/localstack/aws/skeleton.py", line 153, in invoke
    return self.dispatch_request(context, instance)
  File "/opt/code/localstack/localstack/aws/skeleton.py", line 165, in dispatch_request
    result = handler(context, instance) or {}
  File "/opt/code/localstack/localstack/aws/forwarder.py", line 67, in _call
    return fallthrough_handler(context, req)
  File "/opt/code/localstack/localstack/services/moto.py", line 83, in _proxy_moto
    return call_moto(context)
  File "/opt/code/localstack/localstack/services/moto.py", line 46, in call_moto
    return dispatch_to_backend(context, dispatch_to_moto, include_response_metadata)
  File "/opt/code/localstack/localstack/aws/forwarder.py", line 120, in dispatch_to_backend
    http_response = http_request_dispatcher(context)
  File "/opt/code/localstack/localstack/services/moto.py", line 111, in dispatch_to_moto
    response = dispatch(request, request.url, request.headers)
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/core/responses.py", line 225, in dispatch
    return cls()._dispatch(*args, **kwargs)
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/core/responses.py", line 366, in _dispatch
    return self.call_action()
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/core/responses.py", line 455, in call_action
    response = method()
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/logs/responses.py", line 177, in describe_log_groups
    groups, next_token = self.logs_backend.describe_log_groups(
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/utilities/paginator.py", line 57, in pagination_wrapper
    results = func(*args, **kwargs)
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/logs/models.py", line 637, in describe_log_groups
    groups = [
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/logs/models.py", line 638, in <listcomp>
    group.to_describe_dict()
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/logs/models.py", line 449, in to_describe_dict
    "storedBytes": sum(s.stored_bytes for s in self.streams.values()),
  File "/opt/code/localstack/.venv/lib/python3.10/site-packages/moto/logs/models.py", line 449, in <genexpr>
    "storedBytes": sum(s.stored_bytes for s in self.streams.values()),
RuntimeError: dictionary changed size during iteration
```

**Comment by bentsku** [comment]
@ArzelaAscoIi interesting, it seems the issue is with `logs` and `describe_log_groups` and not S3. Are you executing some additional requests than S3?

**Comment by bentsku** [comment]
Okay, so it is working but there's some exceptions. I believe the exceptions could be related to how CloudWatch adds logs for SQS queue metrics, but not certain. Thanks a lot for the report, we will have a look!

**Comment by steffyP** [comment]
Hi @ArzelaAscoIi,

we were not able to reproduce your issue yet, but we tried to fix the race condition you have encountered and provided with the stacktrace. Could you please try to pull `latest` and run your scenario again 🙏 

Please let us know if you still encounter the issue.

**Comment by steffyP** [comment]
Any updates @ArzelaAscoIi?

**Comment by ArzelaAscoIi** [comment]
Sorry for the late reply 🙃 we don't see the errors any more either :) Thanks for the fix!

**Comment by ArzelaAscoIi** [comment]
fixed by upgrading[ localstack 2.0.2 ](https://github.com/localstack/localstack/releases/tag/v2.0.2)
